### DISCUSSION TITLE
Distributed MinIO cluster: Read failed. Insufficient number of drives online

### DISCUSSION THREAD

**Discussion by mmartinello** [minio/minio#17374] [Q&A] [answered]
I'm deploying a distributed MinIO cluster with 4 nodes using Docker.

Every node is running MinIO with the following run command:

```
server --console-address :9001 http://storage{1...4}.my.domain/data
```

MinIO starts correctly on every node but services are not available:

Logs from node `storage1.my.domain`:
```
minio_1  | API: SYSTEM()
minio_1  | Time: 15:14:59 UTC 06/07/2023
minio_1  | Error: Read failed. Insufficient number of drives online (*errors.errorString)
minio_1  |        6: internal/logger/logger.go:258:logger.LogIf()
minio_1  |        5: cmd/prepare-storage.go:246:cmd.connectLoadInitFormats()
minio_1  |        4: cmd/prepare-storage.go:304:cmd.waitForFormatErasure()
minio_1  |        3: cmd/erasure-server-pool.go:103:cmd.newErasureServerPools()
minio_1  |        2: cmd/server-main.go:810:cmd.newObjectLayer()
minio_1  |        1: cmd/server-main.go:611:cmd.serverMain()
minio_1  | Waiting for a minimum of 2 drives to come online (elapsed 18m26s)
minio_1  |
minio_1  | Unable to read 'format.json' from http://storage1.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
minio_1  |
minio_1  | Unable to read 'format.json' from http://storage2.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
minio_1  |
minio_1  | Unable to read 'format.json' from http://storage3.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
minio_1  |
minio_1  | Unable to read 'format.json' from http://storage4.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
```

Logs from node `storage2.my.domain`, `storage3.my.domain` and `storage4.my.domain`:

```
minio_1  |API: SYSTEM()
minio_1  |Time: 14:56:33 UTC 06/07/2023
minio_1  |Error: Marking storage1.my.domain:9000 offline temporarily; caused by Post "http://storage1.my.domain:9000/minio/storage/data/v49/readall?disk-id=&file-path=format.json&volume=.minio.sys": dial tcp 192.168.100.91:9000: connect: connection refused (*fmt.wrapError)
minio_1  |       8: internal/logger/logonce.go:118:logger.(*logOnceType).logOnceIf()
minio_1  |       7: internal/logger/logonce.go:149:logger.LogOnceIf()
minio_1  |       6: internal/rest/client.go:262:rest.(*Client).Call()
minio_1  |       5: cmd/storage-rest-client.go:168:cmd.(*storageRESTClient).call()
minio_1  |       4: cmd/storage-rest-client.go:555:cmd.(*storageRESTClient).ReadAll()
minio_1  |       3: cmd/format-erasure.go:391:cmd.loadFormatErasure()
minio_1  |       2: cmd/format-erasure.go:327:cmd.loadFormatErasureAll.func1()
minio_1  |       1: github.com/minio/pkg@v1.7.2/sync/errgroup/errgroup.go:123:errgroup.(*Group).Go.func1()
ESC[36mminio_1  |ESC[0m
minio_1  |API: SYSTEM()
minio_1  |Time: 14:56:33 UTC 06/07/2023
minio_1  |Error: Read failed. Insufficient number of drives online (*errors.errorString)
minio_1  |       6: internal/logger/logger.go:258:logger.LogIf()
minio_1  |       5: cmd/prepare-storage.go:246:cmd.connectLoadInitFormats()
minio_1  |       4: cmd/prepare-storage.go:286:cmd.waitForFormatErasure()
minio_1  |       3: cmd/erasure-server-pool.go:103:cmd.newErasureServerPools()
minio_1  |       2: cmd/server-main.go:810:cmd.newObjectLayer()
minio_1  |       1: cmd/server-main.go:611:cmd.serverMain()
minio_1  |
minio_1  |Client 'http://storage1.my.domain:9000/minio/storage/data/v49' re-connected in 362.09293ms
minio_1  |Client 'http://storage1.my.domain:9000/minio/storage/data/v49' re-connected in 392.260438ms
minio_1  |
minio_1  |Unable to read 'format.json' from http://storage1.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
minio_1  |
minio_1  |Unable to read 'format.json' from http://storage2.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
minio_1  |
minio_1  |Unable to read 'format.json' from http://storage3.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
minio_1  |
minio_1  |Unable to read 'format.json' from http://storage4.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
```

If I try to access to http://storage1.my.domain:9000/minio/storage/data/v49/readall?disk-id=&file-path=format.json&volume=.minio.sys I get this:

```
curl "http://storage1.my.domain:9000/minio/storage/data/v49/readall?disk-id=&file-path=format.json&volume=.minio.sys"                        Wed Jun  7 15:02:07 2023
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AllAccessDisabled</Code><Message>All access to this resource has been disabled.</Message><Resource>/minio/storage/data/v49/readall</Resource><RequestId></RequestId><HostId></HostId></Error>
```

Every server see each other and firewall rules are permitting access to ports `9000` and `9001` from one node to each others.

Could you help me please?
Thank you!

**Comment by AlexZIX** [comment]
Hi,

1. Check if you running same containers with same versions of MinIO on each node.
2. Check data directory on every node - it should contain folder .minio.sys with format.json inside. If no - check folders permissions. I suppose that you should use folder outside of container for storing data.

**Reply by mmartinello** [reply]
Hi,
1. yes, they are running the same version (I am running the `latest` tag of the Docker container on each node
2. data directory does not contain a `.minio.sys` folder, I've checked permissions and they sounds good, I can create files into the container in the `/data` directory as `root`, and MinIO runs as `root`

I'm continuing to receive `XMinioServerNotInitialized` error from `GET` requests.

**Reply by vadmeste** [reply]
> yes, they are running the same version (I am running the latest tag of the Docker container on each node

Can you set an explicit version instead of 'latest' ?

**Reply by mmartinello** [reply]
I've now tried explicating the version `RELEASE.2023-06-09T07-32-12Z` but I'm having the same problem:

```
<Error>
<Code>XMinioServerNotInitialized</Code>
<Message>Server not initialized, please try again.</Message>
<BucketName>data</BucketName>
<Resource>/data</Resource>
<RequestId>176840D55C9AC379</RequestId>
<HostId>dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8</HostId>
</Error>
```

**Reply by mmartinello** [reply, marked as answer]
I solved the problem setting `network_mode: host`.
Now the cluster works!

**Reply by vadmeste** [reply]
Thanks @mmartinello, are you using docker compose ? if yes, can you share it, I would like to reproduce and fix this message:

```
minio_1  | Unable to read 'format.json' from http://storage1.my.domain:9000/data: Do not upgrade one server at a time - please follow the recommended guidelines mentioned here https://github.com/minio/minio#upgrading-minio for your environment
```

**Reply by mmartinello** [reply]
Yes, I'm using Docker Compose!

Below you can find the `docker-compose.yml` file (basically it's the same on all 4 nodes):

This does not work (with ports mapping):
```
version: '3'

services:
  minio:
    image: minio/minio:latest
    volumes:
      - /data/minio:/data
    ports:
      - 9000:9000
      - 9001:9001
    restart: always
    command: server --address :9000 --console-address :9001 http://storage{1...4}.my.domain:9000/data
```

This works (without ports mapping and with `network_mode: host`)

```
version: '3'

services:
  minio:
    image: minio/minio:latest
    volumes:
      - /data/minio:/data
    restart: always
    command: server --address :9000 --console-address :9001 http://storage{1...4}.my.domain:9000/data
    network_mode: host
```

Thanks!

**Reply by AndrewSav** [reply]
@vadmeste I have the same symptoms and setting `network_mode: host` does resolve this, however, I would like to know how to run it _without_ this settings. The port forwarding should work as expected here, where you able to find out what's going on?

**Reply by harshavardhana** [reply]
Well this is simple you are attempting a DNS host that is resolving to a different IP than where you are trying to run the container on.

Net Host solves this because your container ip now is the host IP.

You should be running the MinIO server inside containers based on container hostnames not public hostnames.

**Reply by AndrewSav** [reply]
@harshavardhana, thank you, I double checked that it is not the case. The hosts (except for the localhost) resolve to the same IP in and out of the container.

> You should be running the MinIO server inside containers based on container hostnames not public hostnames.

If it's a container hostname and not a public hostname, one node won't be able to reach another one, since container hostname is not routable between the two nodes, whereas public hostname is. Can you explain a bit more please?

**Reply by AndrewSav** [reply]
@mmartinello, which operating system you were on? Interestingly this happens to me on Debian, but when I used Ubuntu host instead of Debian I no longer have this issue. Docker version and minio container tag are exactly the same in both cases. If you were on Debian to, then something is going on specific for that distro.

**Comment by egism** [comment]
Having a similar issue. At least the error messages are the same. The command in the container looks like this `server --console-address ":9001" http://102.122.182.50/data http://102.122.182.60/data`. Tried providing hostnames instead of IPs using /etc/hosts, nothing changes. Directory permissions are set to 1001.

Running single node `server --console-address ":9001" /data` works fine. Both cases `/data` dir is mounted in the container.

I am out of ideas how to debug this further. Tried `mc admin trace -a --debug local`. It reveals couple initial requests and responses to `POST /minio/bootstrap/v1/verify` that result to 200 OK.

**Reply by AlexZIX** [reply]
Try this instead:

server --console-address ":9001" http://10.120.180.{5...6}0/data

**Reply by egism** [reply]
Unfortunately nothing changes

**Comment by CatCodeMe** [comment]
> Try this instead:
> 
> server --console-address ":9001" [http://10.120.180.{5...6}0/data](http://10.120.180.%7B5...6%7D0/data)

it works for me ,  but I think this should be optimized. Suppose that I have ip  '1.2.3.4', '6.7.8.9' this mode will not work
